Update Python 3 compatibility. WIP.

janwijbrand opened the pull request.

janjaapdriessen committed:
python3 compatibility

janwijbrand committed:
Revive this Python 3 compat branch.

janwijbrand committed:
merge master back into branch - we did double work here :(

janwijbrand committed:
untrack .coverage

janwijbrand committed:
fixup a typo.

janwijbrand committed:
get tox to run again

janwijbrand committed:
be explicit about the content type of the views.

mgedmin commented on .gitignore:
```diff
@@ -1,13 +1,16 @@
+.tox/
+.tox/
+*.egg-info/
+*.egg-info/
```
`.tox/` and `*.egg-info/` appear twice?

mgedmin commented on .gitignore:
```diff
@@ -1,13 +1,16 @@
-.tox/
-.coverage
 htmlcov/
+parts
```
Why drop the `/` from `parts/`?

mgedmin commented on .travis.yml:
```diff
@@ -2,14 +2,15 @@ language: python
     - pip install -U pip setuptools
     - pip install -U coverage coveralls
-    - pip install -U -e .[test]
+    - pip install -U -e .[test,browser]
```
Perhaps `[test]` should include everything in `[browser]`?

mgedmin replied:
In fact setup.py already makes sure that `[test]` is a superset of `[browser]`.

mgedmin commented on buildout.cfg:
```diff
@@ -31,3 +36,6 @@ domain = zope.errorview
   <include package="zope.i18n" file="meta.zcml" />
+
+[sources]
+z3c.recipe.i18n = git ${buildout:github}/z3c.recipe.i18n pushurl=${buildout:github_push}/z3c.recipe.i18n branch=py3
```
You plan to remove these once a Python-3-capable z3c.recipe.i18n is released to PyPI?

mgedmin commented on src/zope/errorview/browser.py:
```diff
@@ -52,6 +52,8 @@ def update(self):
         self.request.response.setHeader(
             'Pragma', 'no-cache')
+        self.request.response.setHeader(
+            'Content-Type', 'text/plain;charset=utf-8')
```
This doesn't seem to be related to Python 3 changes?

janwijbrand replied:
This relates to one of your general remarks above.

These changes are in fact related to Python 3. Since in Python 2 when returning "just" a string as a response would have `zope.publisher` not require a content-type to be set as the response body would not be unicode.

In Python 3 "just" returning a string as response obviously *is* unicode and `zope.publisher` then requires (otherwise https://github.com/zopefoundation/zope.publisher/blob/c1ba72f6a3838269badc433ba9f06a4f1a0eabf2/src/zope/publisher/http.py#L821 would fail) a content type to be set. In other words, it became more important than before to make sure you set the correct content type in your application for your views - which in itself is a good thing I think.

I'll make a note of this in CHANGES.rst indeed and bump the version to 1.1.

mgedmin replied:
Thank you for explaining!

mgedmin commented on src/zope/errorview/http.py:
```diff
@@ -40,6 +40,8 @@ def __init__(self, context, request):
     def update(self):
         self.request.response.setStatus(500)
+        self.request.response.setHeader(
+            'Content-Type', 'text/plain; charset=utf-8')
```
And this.

janwijbrand replied:
Idem.

gyst replied:
This has a spurious extra space.

mgedmin replied:
Which space is spurious?

gyst replied:
Seems curious that the request header in `browser` should *not* have a space after the semicolon, while the response header here does add that space.

mgedmin reviewed (commented):
Mostly LGTM.

Missing bits:

- setup.py updates
- changelog

Things I'm worried about:

- Content-Type change to text/plain. If people are subclassing zope.errorpage views and rendering HTML, their sites will break.  This may be fine, if we bump the major version number and put a prominent notice in the changelog.

janwijbrand committed:
z3c.recipe.i18n was released.

janwijbrand committed:
remove duplicates and z3c.recipe.i18n was released.

processed comments https://github.com/zopefoundation/zope.errorview/pull/6#discussion_r160472248 and https://github.com/zopefoundation/zope.errorview/pull/6#discussion_r160472292 and https://github.com/zopefoundation/zope.errorview/pull/6#discussion_r160472550

janwijbrand committed:
Update changelog making a note about the content-type header that now is set.

janwijbrand commented:
@mgedmin thanks for the review. I think I processed all comments. If you're satisfied I'll make a release (unblocking further work on groktoolkit).

mgedmin commented on buildout.cfg:
```diff
@@ -7,9 +7,9 @@ parts =
-  z3c.recipe.i18n
+
+[versions]
+z3c.recipe.i18n = 1.0.0
```
I'm not too keen on pinning exact versions: we want to know if newer z3c.recipe.i18n releases somehow break us.

If you need to specify a minimum, buildout lets you say `zc.recipe.i18n = >= 1.0.0`.

janwijbrand replied:
will do.

mgedmin reviewed (approved):
LGTM with one caveat, but I don't think that's a blocker.

janwijbrand committed:
use *at least* 1.0.0
